File: internal/analyzer/detectors/port_scanner.py
from typing import Dict, List, Set

import pandas as pd

from internal.analyzer.base import Alert, BaseDetector, Level


class PortScannerDetector(BaseDetector):
    WHITELIST_IPS = {"192.168.0.1", "127.0.0.1"}
    IGNORE_PORTS = {1900, 5353, 3702}

    def __init__(self, port_threshold: int = 20):
        super().__init__()
        self.port_threshold = port_threshold

    def analyze(self, df: pd.DataFrame) -> List[Alert]:
        alerts = []
        if df.empty or "src_ip" not in df.columns or "dst_port" not in df.columns:
            return alerts

        clean_df = df.dropna(subset=["src_ip", "dst_port"])
        clean_df = clean_df[
            (clean_df["src_ip"].str.strip() != "") & (clean_df["dst_port"] != 0)
        ]

        if clean_df.empty:
            return alerts

        clean_df = clean_df[~clean_df["src_ip"].isin(self.WHITELIST_IPS)]
        clean_df = clean_df[~clean_df["dst_port"].isin(self.IGNORE_PORTS)]

        if clean_df.empty:
            return alerts

        stats = clean_df.groupby("src_ip")["dst_port"].nunique()
        scanners = stats[stats >= self.port_threshold].index.tolist()

        for ip in scanners:
            port_count = int(stats[ip])
            latest_timestamp = clean_df[clean_df["src_ip"] == ip]["timestamp"].max()

            if self._should_alert("PORT_SCAN", ip):
                alerts.append(
                    Alert(
                        timestamp=str(latest_timestamp),
                        level=Level.WARNING,
                        type="PORT_SCAN",
                        message=f"IP {ip} is scanning ports: {port_count} unique ports hit.",
                        src_ip=ip,
                    )
                )

        return alerts

File: internal/analyzer/detectors/port_scanner.go
package detectors

import (
	"fmt"
	"hattic/internal/analyzer"
	"sync"
	"time"
)

type PortScannerDetector struct {
	*analyzer.BaseDetector
	portThreshold int
	ipToPorts     map[string]map[uint16]bool
	mu            sync.Mutex
	whitelistIPs  map[string]bool
	ignorePorts   map[uint16]bool
}

func NewPortScannerDetector(threshold int) *PortScannerDetector {
	return &PortScannerDetector{
		BaseDetector:  analyzer.NewBaseDetector(5 * time.Minute),
		portThreshold: threshold,
		ipToPorts:     make(map[string]map[uint16]bool),
		whitelistIPs: map[string]bool{
			"192.168.0.1": true,
			"127.0.0.1":   true,
		},
		ignorePorts: map[uint16]bool{
			1900: true,
			5353: true,
			3702: true,
		},
	}
}

func (d *PortScannerDetector) Analyze(packet *analyzer.PacketInfo) []analyzer.Alert {
	var alerts []analyzer.Alert

	if packet.SrcIP == "" || packet.DstPort == 0 {
		return alerts
	}

	if d.whitelistIPs[packet.SrcIP] || d.ignorePorts[packet.DstPort] {
		return alerts
	}

	d.mu.Lock()
	defer d.mu.Unlock()

	if _, exists := d.ipToPorts[packet.SrcIP]; !exists {
		d.ipToPorts[packet.SrcIP] = make(map[uint16]bool)
	}

	d.ipToPorts[packet.SrcIP][packet.DstPort] = true
	portCount := len(d.ipToPorts[packet.SrcIP])

	if portCount >= d.portThreshold {
		if d.ShouldAlert("PORT_SCAN", packet.SrcIP) {
			alerts = append(alerts, analyzer.Alert{
				Timestamp: packet.Timestamp.Format(time.RFC3339),
				Level:     analyzer.LevelWarning,
				Type:      "PORT_SCAN",
				Message:   fmt.Sprintf("IP %s is scanning ports: %d unique ports hit.", packet.SrcIP, portCount),
				SrcIP:     packet.SrcIP,
			})
		}
	}

	return alerts
}
